extension/src/SymbolReference/BinaryReader.ts

```diff
@@ -0,0 +1,74 @@
+export class BinaryReader {
+  public byteLength: number;
+  public byteOffset: number;
+  public littleEndian: boolean;
+
+  private buffer: Buffer;
+  private isArrayBuffer: boolean;
+
+  constructor(buffer: Buffer, littleEndian: boolean) {
+    this.buffer = buffer;
+    this.byteLength = buffer.byteLength;
+    this.byteOffset = buffer.byteOffset;
+    this.isArrayBuffer = buffer instanceof ArrayBuffer;
+    this.littleEndian = littleEndian;
+  }
+
+  private checkBounds(byteOffset: number, byteLength: number): void {
+    if (byteLength < 0) {
+      throw new RangeError("Length is negative.");
+    }
+    if (byteOffset < 0 || byteOffset + byteLength > this.byteLength) {
+      throw new RangeError("Offsets are out of bounds.");
+    }
+  }
+
+  getBytes(length: number, byteOffset: number): Uint8Array {
+    this.checkBounds(byteOffset, length);
+
+    byteOffset += this.byteOffset;
+
+    const result = this.isArrayBuffer
+      ? new Uint8Array(this.buffer, byteOffset, length)
+      : (this.buffer.slice || Array.prototype.slice).call(
+          this.buffer,
+          byteOffset,
+          byteOffset + length
+        );
+
+    return this.littleEndian || length <= 1 ? result : result.reverse();
+  }
+
+  getUint32(byteOffset: number): number {
+    const b = this.getBytes(4, byteOffset);
+    return (b[3] << 24) | (b[2] << 16) | (b[1] << 8) | b[0];
+  }
+
+  getUint16(byteOffset: number): number {
+    const b = this.getBytes(2, byteOffset);
+    return (b[1] << 8) | b[0];
+  }
+
+  getUint64(byteOffset: number): Uint64 {
+    const parts = this.littleEndian ? [0, 4] : [4, 0];
+
+    for (let i = 0; i < 2; i++) {
+      parts[i] = this.getUint32(byteOffset + parts[i]);
+    }
+
+    return new Uint64(parts[0], parts[1]);
+  }
+}
+
+class Uint64 {
+  constructor(public lo: number, public hi: number) {}
+  valueOf(): number {
+    return this.lo + 4294967296 * this.hi;
```

Comment:
4294967296 might earn its own constant?

Reply:
yeah, name suggestion? :D

Reply:
UINT_MAX

Reply:
the hardest thing for developers, naming... 😁

Or `UINT32_MAX`?